core/storage/statistics/gae_models.py

```diff
@@ -737,3 +845,247 @@ def _get_entity_id(cls, exploration_id, exploration_version, state_name,
+                in model.started_large_answer_bucket_ids):
+            raise Exception(
+                'Expected to not have started migrating large answer bucket: '
+                '\'%s\' (part of group \'%s\')' % (
+                    large_answer_bucket_id, state_answer_log_model_item_id))
+        if not model:
+            started_large_answer_bucket_ids = (
+                [large_answer_bucket_id] if large_answer_bucket_id else [])
+            model = cls(
+                id=state_answer_log_model_item_id,
+                exploration_id=exploration_id, state_name=state_name,
+                exploration_versions=[], started_migration=True,
+                started_large_answer_bucket_ids=started_large_answer_bucket_ids,
+                finished_large_answer_bucket_ids=[],
+                expected_large_answer_bucket_count=large_answer_bucket_count)
+        else:
+            started_large_answer_bucket_ids = set(
+                model.started_large_answer_bucket_ids)
+            started_large_answer_bucket_ids.add(large_answer_bucket_id)
+            model.started_large_answer_bucket_ids = list(
+                started_large_answer_bucket_ids)
+        model.put()
+
+    @classmethod
+    def start_migrating_answer_bucket(
+            cls, state_answer_log_model_item_id, exploration_id, state_name,
+            large_answer_bucket_id, large_answer_bucket_count):
+        transaction_services.run_in_transaction(
+            cls._start_migrating_answer_bucket, state_answer_log_model_item_id,
+            exploration_id, state_name, large_answer_bucket_id,
+            large_answer_bucket_count)
+
+    @classmethod
+    def _finish_migrating_answer(
+            cls, state_answer_log_model_item_id, exploration_version):
+        model = cls.get(state_answer_log_model_item_id)
+        # Avoid an unnecessary put if the given version is already recorded
+        if exploration_version not in model.exploration_versions:
+            model.exploration_versions.append(exploration_version)
+            model.put()
+
+    @classmethod
+    def finish_migrating_answer(
+            cls, state_answer_log_model_item_id, exploration_version):
+        transaction_services.run_in_transaction(
+            cls._finish_migrating_answer, state_answer_log_model_item_id,
+            exploration_version)
+
+    @classmethod
+    def _finish_migration_answer_bucket(
+            cls, state_answer_log_model_item_id, large_answer_bucket_id):
+        model = cls.get(state_answer_log_model_item_id)
+        if large_answer_bucket_id:
+            finished_large_answer_bucket_ids = set(
+                model.finished_large_answer_bucket_ids)
+            finished_large_answer_bucket_ids.add(large_answer_bucket_id)
+            expected_bucket_count = model.expected_large_answer_bucket_count
+            model.finished_migration = (
+                len(finished_large_answer_bucket_ids) >= expected_bucket_count)
```

Comment:
Is the >= for safety, or is there actually a case where > can legitimately occur?